qualcoder/view_graph.py

```diff
@@ -44,6 +44,101 @@
 path = os.path.abspath(os.path.dirname(__file__))
 logger = logging.getLogger(__name__)
 
+def get_cats_from_supercats(node,cats):
+    cats = deepcopy(cats)
+    for x in cats:
+        if x['supercatid'] == node['catid']:
+            yield x
+            yield from get_cats_from_supercats(x,cats)
+
+def calc_supercats(cats):
+    per_supercat = defaultdict(list)
+    for cat in cats:
+        per_supercat[cat['supercatid']].append(cat)
+    return per_supercat
+
+def recurse_supercats(node,per_supercat,func):
+    for x in per_supercat.pop(node['catid'],[]):
+        func(node,x)
+        yield x
+        yield from recurse_supercats(x,per_supercat,func=func)
+
+def visit_cats_from_supercats(cats,node=None,func=None):
+    if func is None:
+        func = lambda a,b:(a,b)
+    per_supercat = calc_supercats(cats)
+    if node is not None:
+        yield from recurse_supercats(node,per_supercat,func=func)
+    else:
+        per_cats = {x['catid']:x for x in cats}
+        for supercatid,scats in per_supercat.items():
+            if supercatid in per_cats:
+                supercat = per_cats[supercatid]
+                for cat in scats:
+                    func(supercat,cat)
+                    yield cat
+            else:
+                for cat in cats:
+                    func(None,cat)
+                    yield cat
+
+
+def get_codes_from_cats(cats,codes):
+    for cat in cats:
+        for c in codes:
+            if c['catid'] == cat['catid']:
+                yield c
+
+def visit_codes_from_cats(cats,codes,func=None):
```

Comment:
OK I think I get these sections of codes to create the tree. Would be good to add some comments for methods: get codes from cats, visit codes from cats.

Reply:
i agree, i will try to add these